docs/content/en/highlights/2021-07-13-helm-customConfig.md

```diff
@@ -0,0 +1,324 @@
+     sinkId: vector_sink
+     inputs: ["kubernetes_logs"]
+     host: vector
+     port: "9000"
+     config: {}
+     rawConfig: null
+   ...
++  customConfig:
++    ...
++    sinks:
++      ...
++      vector_sink:
++        type: vector
++        inputs: ["kubernetes_logs"]
++        address: vector:9000
++    ...
+```
+
+### internalMetricsSource
+
+```diff title="values.yaml"
+   internalMetricsSource:
++    enabled: false
+     sourceId: internal_metrics
+     config: {}
+     rawConfig: null
+   ...
++  customConfig:
++    ...
++    sources:
++      ...
++      internal_metrics:
++        type: internal_metrics
++    ...
+```
+
+### hostMetricsSource
+
+Users should set the PROCFS_ROOT and SYSFS_ROOT environment variables, as well as mounting the
+required hostPaths.
+
+```diff title="values.yaml"
+   hostMetricsSource:
++    enabled: false
+     sourceId: host_metrics
+     config:
+       filesystem:
+         devices:
+           excludes: [binfmt_misc]
+         filesystems:
+           excludes: [binfmt_misc]
+         mountpoints:
+           excludes: ["*/proc/sys/fs/binfmt_misc"]
+     rawConfig: null
+   ...
++  customConfig:
++    ...
++    sources:
++      ...
++      host_metrics:
++        type: host_metrics
++        filesystem:
++          devices:
++            excludes: ["binfmt_misc"]
++          filesystems:
++            excludes: ["binfmt_misc"]
++          mountPoints:
++            excludes: ["*/proc/sys/fs/binfmt_misc"]
++  env:
++    - name: PROCFS_ROOT
++      value: /host/proc
++    - name: SYSFS_ROOT
++      value: /host/sys
++  extraVolumeMounts:
++    - name: procfs
++      mountPath: /host/proc
++      readOnly: true
++    - name: sysfs
++      mountPath: /host/sys
++      readOnly: true
++  extraVolumes:
++    - name: procfs
++      hostPath:
++        path: /proc
++    - name: sysfs
++      hostPath:
++        path: /sys
++    ...
+```
+
+### prometheusSink
+
+The `extraContainerPorts` or `service` key should be used to expose the port configured in `customConfig`.
+
+The `prometheusSink.podMonitor` key has been moved to a top level key and can be accessed directly at
+`podMonitor`. The `addPodAnnotations` option has been removed in favor setting the required annotations
+with the `podAnnotations` key.
+
+```diff title="values.yaml"
+   prometheusSink:
++    enabled: false
+     sinkId: prometheus_sink
+     inputs: []
+     excludeInternalMetrics: false
+     listenAddress: "0.0.0.0"
+     listenPort: "9090"
+     config: {}
+     rawConfig: null
+   ...
++  customConfig:
++    ...
++    sinks:
++      ...
++      prometheus_sink:
++        type: prometheus_exporter
++        inputs: ["host_metrics", "internal_metrics"]
++        address: 0.0.0.0:9090
++    ...
+```
+
+## Using `customConfig`
+
+We've configured the ConfigMap template to `fail` if the deprecated keys are enabled at the same time
+as the new YAML based configurations. The following values can be used to disable the old deprecated
```

Comment:
Instead of requiring users to opt-out of the deprecated keys, can't we make the `customConfig` and the deprecated keys mutually exclusive. By providing the `customConfig` option we disable the deprecated keys and require that the user remove them from their `cusomConfig.yaml`.

Reply:
I mostly wanted to ensure it was obvious that the deprecated configurations wouldn't be merged or combined with anything provided in `customConfig` - but if that's not entirely necessary it's easy enough to pick _just_ customConfig over any of the now deprecated keys.

Reply:
Am I missing something or is:

> Instead of requiring users to opt-out of the deprecated keys, can't we make the `customConfig` and the deprecated keys mutually exclusive. By providing the `customConfig` option we disable the deprecated keys and require that the user remove them from their `cusomConfig.yaml`.

not what is already happening? If the user specifies both the deprecated keys and `configMap`, the template fails, no?

Reply:
However, I can see that the current keys still default to enabled. Would it be possible to have them automatically disable themselves if they are not specified in the config file by the user and `customConfig` is specified? That would make the opting into the new strategy a bit more painless.

Reply:
The way I read @binarylogic's suggestion was to ignore the deprecated keys if the new keys are used, rather than erroring - might be wrong though!

Reply:
I think my preference would be:

* If user has deprecated keys explicitly specified, and then specifies `customConfig`: error
* If the user does not have deprecated keys explicitly specified, and then specifies `customConfig`: default deprecated keys to `enabled: false`

Though, I'm not sure if that is even possible.

As a second choice, I like what you have: just ignore the deprecated keys when `customConfig` is specified.